_posts/2019-05-21-accessing-kafka-part-5.md

```diff
@@ -0,0 +1,218 @@
+
+Although some Ingress controllers also support working directly with TCP connections, TLS passthrough seems to be more widely supported.
+Therefore we decided to prefer TLS passthrough in Strimzi.
+That also means that when using Ingress, TLS encryption will always be enabled.
+
+One of the main differences compared to OpenShift Routes is that for Ingress you have to specify the host address in your Kafka custom resource.
+The Router in OpenShift will automatically assign a host address based on the route name and the project.
+But in Ingress the host address has to be specified in the Ingress resource.
+You also have to take care that DNS resolves the host address to the ingress controller.
+Strimzi cannot generate it for you, because it does not know which DNS addresses are configured for the Ingress controller.
+
+If you want to try it for example on Minikube or in other environments where you don't have any managed DNS service to add the hosts for the Kafka cluster, you can use one of the wildcard DNS services such as [nip.io](https://nip.io/) or [xip.io](http://xip.io/) and set it to point to the IP address of your Ingress controller.
+For example `broker-0.<minikube-ip-address>.nip.io`.
+
+![Kafka clients connecting through Ingress controller]({{ "/assets/2019-05-21-connecting-through-ingress" }})
+
+The way Strimzi uses Ingress to expose Kafka should already be familiar to you from the previous blog posts.
+We create one service as a bootstrap service and additional services for individual access to each of the Kafka brokers in your cluster.
+For each of these services we will also create one Ingress resource with the corresponding TLS passthrough rule.
+
+![Accessing Kafka using Ingress]({{ "/assets/2019-05-21-ingress-access.png" }})
+
+When configuring Strimzi to use Ingress, you have to specify the type of the external listener as `ingress` and specify the ingress hosts used for the different brokers as well as for bootstrap in the `configuration` field:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta1
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+    listeners:
+      # ...
+      external:
+        type: ingress
+        configuration:
+          bootstrap:
+            host: bootstrap.192.168.64.46.nip.io
+          brokers:
+          - broker: 0
+            host: broker-0.192.168.64.46.nip.io
+          - broker: 1
+            host: broker-1.192.168.64.46.nip.io
+          - broker: 2
+            host: broker-2.192.168.64.46.nip.io
+    # ...
+```
+
+Using Ingress in your clients is very similar to openShift Routes.
+Since it always uses TLS Encryption, you have to first download the server certificate (replace `my-cluster` with the name of your cluster):
+
+```
+oc extract secret/my-cluster-cluster-ca-cert --keys=ca.crt --to=- > ca.crt
+keytool -import -trustcacerts -alias root -file ca.crt -keystore truststore.jks -storepass password -noprompt
+```
```

Comment:
Because Ingress is a thing related to Kubernetes as Routes to OpenShift shouldn't we use the kubectl way to get the certificate? And not oc tool?

Reply:
Good point. Its shame `kubectl` doesn't have the `extract` command.